extend_foreach_to_cover_generate_rules.md

```diff
@@ -0,0 +1,182 @@
+
+Examples:
+1. Generating a list of new network policies based on a comma separated string in an annotation using a `forEach` and `Split()`.
+2. Migrating from ingress to Kubernetes Gateway API requires generating at least two resources: `Gateway` and `HTTPRoute`.
+
+# Proposal
+In this proposal, Kyverno generate rule can be extended to allowing the generation of multiple resources using `forEach`.
+This could be done by using the `list` attribute which is a JMESPath expression that defines the sub-elements it processes.
+
+Examples:
+1. Generate multiple network policies when a new namespace is created:
+```yaml
+spec:
+  rules:
+  - name: generate-network-policies
+    match:
+      any:
+      - resources:
+          kinds:
+          - Namespace
+    generate:
+      synchronize: true
+      foreach:
+        - list: request.object.metadata.labels.networkpolicies | split(@, ',')
+          resources:
+            - apiVersion: v1
+              kind: NetworkPolicy
+              name: '{{ element }}'
+              namespace: "{{request.object.metadata.name}}"
+              data:
+                spec:
+                  podSelector: {}
+                  policyTypes:
+                  - Ingress
+                  - Egress
+```
+
+2. Generate ConfigMap and Secret for each container:
+```yaml
+spec:
+  rules:
+  - name: generate-configmaps-and-secrets
+    match:
+      any:
+      - resources:
+          kinds:
+          - Pod
+    generate:
+      synchronize: true
+      foreach:
+        - list: request.object.spec.containers
+          resources:
+            - apiVersion: v1
+              kind: ConfigMap
+              name: cm
+              namespace: "{{request.object.metadata.namespace}}"
+              data:
+                foo: bar
+            - apiVersion: v1
+              kind: Secret
+              name: secret
+              namespace: "{{request.object.metadata.namespace}}"
+              data:
+                foo: bar
+```
+
+3. Generate Gateway and HTTPRoute from Ingress:
+```yaml
+spec:
+  rules:
+  - name: generate-gatewayapi
+    match:
+      any:
+      - resources:
+          kinds:
+          - Ingress
+    generate:
+      synchronize: false
+      foreach:
+        - list: request.object.spec.rules
+          resources:
+            - apiVersion: gateway.networking.k8s.io/v1beta1
+              kind: HTTPRoute
+              # the name must be unique
+              name: httproute
+              namespace: "{{request.object.metadata.namespace}}"
+              data:
+                spec:
+                  hostnames: 
+                    - '{{ element.host }}'
+        - list: request.object.spec.ingressClassName
+          resources:
+            - apiVersion: gateway.networking.k8s.io/v1beta1
+              kind: Gateway
+              name: gateway
+              data:
+                spec:
+                  gatewayClassName: '{{ element }}'
```

Comment:
Currently the mapping for a generate rule is - one rule can support M `foreach`, one `foreach` can generate N resources. Worst case one generate rule manages M*N resources, I wonder the complexity of the management work can be, for example, to sync changes from the rule to downstream resources.

Reply:
For simplicity, we can limit the rule to have 1 `foreach` and it will only generate 1 resource. I amn't sure how we sync the changes given there is M*N managed resources rather than applying the rule again.